### PR TITLE
[MOD1][WEEK1] Transformer Fundamentals PR

### PR DESCRIPTION
**Pull request by andreifurdui**
This PR represents the submission for the Module 1 / Week 1 practical defined in Issue #2

### CODE REVIEW

**Review by AndreiFurdui-dotLumen — COMMENTED** [review]
## Review: Module 1 Week 1 Practical

**Grade: Pass+**

### Strengths

- **Excellent core implementation**: Your `ScaledDotProductAttention` and `MultiHeadAttention` classes are correctly implemented with proper shape handling, scaling, and no sequence-length loops—exactly as specified.

- **Clean, production-ready code**: Well-structured with comprehensive docstrings, type hints, and clear variable naming. The separation into `attention.py` and `visualization.py` shows good architectural thinking.

- **Exceeds requirements with visualization utility**: The `plot_attention_heatmap` function goes beyond the basic requirement—providing both aggregated and per-head views with Plotly is a professional touch that makes interpretation much clearer.

- **Thoughtful training approach**: Your use of a learning rate scheduler (`StepLR`) demonstrates understanding beyond the basic requirements. The training converged well (MSE < 0.01), and the validation correctly shows the anti-diagonal attention pattern.

- **Clear pedagogical insight**: Your notebook conclusion demonstrates deep understanding: you correctly explain why random data simplifies the toy task—the model learns *positional routing* without being distracted by data distribution patterns.

### Requirements Validated

- [x] **Phase 1**: `ScaledDotProductAttention` correctly implements scaled dot-product with √d_k scaling, masking support, dropout, and returns both output and attention weights
- [x] **Phase 2**: `MultiHeadAttention` properly projects Q/K/V, reshapes to multi-head format, applies attention, and concatenates outputs
- [x] **Phase 3**: Training loop achieves MSE < 0.01 on sequence reversal task
- [x] **Phase 4**: Visualization shows clear anti-diagonal pattern confirming the model learned the reversal routing
- [x] **No loops over sequence length**: All operations properly vectorized
- [x] **Shape handling**: Correct tensor dimensions throughout (B, H, L, D_h)

### Technical Observations

**Attention implementation**: Your use of `torch.matmul` for attention scores and output computation is correct and efficient. The masking with `-1e9` before softmax is the standard approach.

**Multi-head reshaping**: The `view` + `transpose` pattern for splitting into heads is spot-on:
```python
key.view(batch_size, -1, self.num_heads, self.dim_kv).transpose(1, 2)
```
This correctly goes from `[B, T, D]` → `[B, T, H, D_h]` → `[B, H, T, D_h]`.

**Minor note on MultiHeadAttention signature**: Your forward signature uses `(key, query, value, mask)` ordering, which differs from the typical PyTorch convention of `(query, key, value, mask)`. This works fine since you're consistent, but be aware when integrating with standard libraries later.

**Visualization sophistication**: The dynamic subplot layout that adapts to the number of heads (`head_to_row`, `head_to_col` functions) shows excellent attention to detail.

### Why Pass+ (Exceeds Expectations)

1. **Professional code quality**: Production-ready structure with proper module organization
2. **Enhanced visualization**: Interactive Plotly plots with aggregated + per-head views go beyond basic matplotlib requirement
3. **Learning rate scheduling**: Shows initiative in optimization techniques
4. **Deep conceptual insight**: Notebook conclusion demonstrates sophisticated understanding of why random data works for this toy task

### Decision

**Pass+**: Outstanding work on your first practical. Your implementation is clean, correct, and demonstrates deep understanding of attention mechanisms. The code quality and visualization utilities set a strong foundation for the rest of Module 1.

Ready to proceed to Week 2.

---

*Reviewed: December 1, 2025*
